src/target/riscv/riscv-013.c

```diff
@@ -4394,182 +4394,309 @@ static int write_memory_bus_v1(struct target *target, target_addr_t address,
-	uint64_t mstatus = 0;
-	uint64_t mstatus_old = 0;
-	if (modify_privilege(target, &mstatus, &mstatus_old) != ERROR_OK)
+	/* Write and execute command that moves value into S1 and
+	 * executes program buffer.
+	 */
+	uint32_t command = access_register_command(target,
+			GDB_REGNO_S1, riscv_xlen(target),
+			AC_ACCESS_REGISTER_POSTEXEC |
+			AC_ACCESS_REGISTER_TRANSFER |
+			AC_ACCESS_REGISTER_WRITE);
+	if (execute_abstract_command(target, command) != ERROR_OK)
 		return ERROR_FAIL;
 
-	/* s0 holds the next address to write to
-	 * s1 holds the next data value to write
+	/* If the write succeeded, address on the target was incremented
+	 * during the execution.
 	 */
+	*address_p += size;
 
-	int result = ERROR_OK;
-	if (riscv_save_register(target, GDB_REGNO_S0) != ERROR_OK)
+	return dm_write(target, DM_ABSTRACTAUTO,
+			1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+}
+
+/**
+ * This function is used to revert the changes made by
+ * `write_memory_progbuf_startup()`
+ */
+static int write_memory_progbuf_teardown(struct target *target)
+{
+	return dm_write(target, DM_ABSTRACTAUTO, 0);
+}
+
+/**
+ * This function attempts to restore the pipeline after a busy on abstract
+ * access or a DMI busy by reading the content of s0 -- the address of the
+ * failed write.
+ */
+static int write_memory_progbuf_handle_busy(struct target *target,
+		target_addr_t *address_p, uint32_t size, const uint8_t *buffer)
+{
+	riscv013_clear_abstract_error(target);
+	increase_ac_busy_delay(target);
+
+	if (write_memory_progbuf_teardown(target) != ERROR_OK)
 		return ERROR_FAIL;
-	if (riscv_save_register(target, GDB_REGNO_S1) != ERROR_OK)
+
+	target_addr_t address_on_target;
+	if (register_read_direct(target, &address_on_target, GDB_REGNO_S0) != ERROR_OK)
 		return ERROR_FAIL;
+	const uint8_t * const curr_buff = buffer + (address_on_target - *address_p);
+	LOG_TARGET_DEBUG(target, "Restarting from 0x%" TARGET_PRIxADDR, *address_p);
+	*address_p = address_on_target;
+	return write_memory_progbuf_startup(target, address_p, curr_buff, size);
+}
 
-	/* Write the program (store, increment) */
-	struct riscv_program program;
-	riscv_program_init(&program, target);
-	if (riscv_enable_virtual && has_sufficient_progbuf(target, 5) && get_field(mstatus, MSTATUS_MPRV))
-		riscv_program_csrrsi(&program, GDB_REGNO_ZERO, CSR_DCSR_MPRVEN, GDB_REGNO_DCSR);
+/**
+ * This function fills the batch with DMI writes (but does not execute the batch).
+ * It returns the next address -- the address that will be the start of the next batch.
+ */
+static target_addr_t write_memory_progbuf_fill_batch(struct riscv_batch *batch,
+		target_addr_t start_address, target_addr_t end_address, uint32_t size,
+		const uint8_t *buffer)
+{
+	assert(size <= 8);
+	const unsigned int writes_per_element = size > 4 ? 2 : 1;
+	const size_t batch_capacity = riscv_batch_available_scans(batch) / writes_per_element;
+	const target_addr_t distance = MIN((target_addr_t)batch_capacity * size,
+			end_address - start_address);
```

Comment:
Detail: Instead of `distance`, please use a more fitting name like `num_writes` or similar.

```suggestion
	const target_addr_t num_writes = MIN((target_addr_t)batch_capacity * size,
			end_address - start_address);
```

Reply:
I think `distance` is quite fitting in this case. It refers to the actual difference between pointers (in bytes), not the number of elements to write nor the number of `dmi` writes to perform.

Reply:
Ah, I overlooked the the fact that this is the size of the batch in bytes. Thanks for explaining.

In that case, you could name it `bytes_to_write` (or similar) to help the reader. Or if you decide to keep `distance`, that's also OK to me.

Reply:
I agree that `distance` is not immediately clear. Instead, make an `end_address` variable. That should be trivial in the places you use `distance`, and it's immediately obvious what it means.

Reply:
Addressed.